File: go-backend-interview-prep/04_interfaces_and_generics/05_errors_as_interfaces/main.py
# Python equivalent of Errors as Interfaces (compare with main.go)
# Go error interface ~ Python Exception hierarchy
# Go errors.Is ~ Python isinstance or direct comparison
# Go errors.As ~ Python except with specific exception type
# Go error wrapping ~ Python exception chaining (raise ... from ...)


# --- Custom exception types ---
class NotFoundError(Exception):
    """Sentinel-style exception."""
    pass


class ValidationError(Exception):
    """Structured exception with extra fields."""
    def __init__(self, field: str, message: str):
        self.field = field
        self.message = message
        super().__init__(f"validation error: {field} - {message}")


def find_user(id: int) -> str:
    """Raises instead of returning (value, error) tuple."""
    if id <= 0:
        raise NotFoundError("not found")
    if id > 100:
        raise ValidationError(field="id", message="must be <= 100")
    return "Alice"


def fetch_profile(id: int) -> str:
    """Wraps exceptions with chaining (like Go's %w)."""
    try:
        name = find_user(id)
        return f"Profile: {name}"
    except Exception as e:
        raise RuntimeError(f"fetch_profile(id={id})") from e


def main():
    # --- Basic error handling ---
    print("--- Basic error handling ---")
    try:
        name = find_user(1)
        print("found:", name)
    except Exception as e:
        print("error:", e)

    # --- Sentinel-style exception ---
    print("\n--- Sentinel exception ---")
    try:
        find_user(0)
    except NotFoundError:
        print("user not found (specific exception match)")

    # --- Custom exception type ---
    print("\n--- Custom exception type ---")
    try:
        find_user(200)
    except ValidationError as e:
        print(f"validation: field={e.field} msg={e.message}")

    # --- Exception chaining (like Go error wrapping) ---
    print("\n--- Exception chaining ---")
    try:
        fetch_profile(0)
    except RuntimeError as e:
        print("wrapped:", e)
        # __cause__ is the chained exception (like Go's Unwrap)
        if isinstance(e.__cause__, NotFoundError):
            print("found NotFoundError in chain")

    # --- Wrapped custom exception ---
    try:
        fetch_profile(200)
    except RuntimeError as e:
        if isinstance(e.__cause__, ValidationError):
            print("found ValidationError in chain:", e.__cause__.field)


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/04_interfaces_and_generics/05_errors_as_interfaces/main.go
package main

import (
	"errors"
	"fmt"
)

// --- Sentinel error ---
var ErrNotFound = errors.New("not found")

// --- Custom error type ---
type ValidationError struct {
	Field   string
	Message string
}

func (e *ValidationError) Error() string {
	return fmt.Sprintf("validation error: %s - %s", e.Field, e.Message)
}

// --- Functions returning errors ---
func findUser(id int) (string, error) {
	if id <= 0 {
		return "", ErrNotFound
	}
	if id > 100 {
		return "", &ValidationError{Field: "id", Message: "must be <= 100"}
	}
	return "Alice", nil
}

func fetchProfile(id int) (string, error) {
	name, err := findUser(id)
	if err != nil {
		// Wrap with %w to preserve the error chain
		return "", fmt.Errorf("fetchProfile(id=%d): %w", id, err)
	}
	return fmt.Sprintf("Profile: %s", name), nil
}

func main() {
	// --- Basic error handling ---
	fmt.Println("--- Basic error handling ---")
	name, err := findUser(1)
	if err != nil {
		fmt.Println("error:", err)
	} else {
		fmt.Println("found:", name)
	}

	// --- Sentinel error ---
	fmt.Println("\n--- Sentinel error ---")
	_, err = findUser(0)
	if errors.Is(err, ErrNotFound) {
		fmt.Println("user not found (sentinel match)")
	}

	// --- Custom error type ---
	fmt.Println("\n--- Custom error type ---")
	_, err = findUser(200)
	var valErr *ValidationError
	if errors.As(err, &valErr) {
		fmt.Printf("validation: field=%s msg=%s\n", valErr.Field, valErr.Message)
	}

	// --- Wrapped error chain ---
	fmt.Println("\n--- Error wrapping ---")
	_, err = fetchProfile(0)
	fmt.Println("wrapped:", err)

	// errors.Is unwraps the chain
	if errors.Is(err, ErrNotFound) {
		fmt.Println("errors.Is found ErrNotFound in the chain")
	}

	// --- Wrapped custom error ---
	_, err = fetchProfile(200)
	if errors.As(err, &valErr) {
		fmt.Println("errors.As found ValidationError in chain:", valErr.Field)
	}

	// --- error is just an interface ---
	fmt.Println("\n--- error is an interface ---")
	var e error = &ValidationError{Field: "email", Message: "required"}
	fmt.Println("error interface:", e)
	fmt.Printf("concrete type: %T\n", e)
}
